doc/nextml2015/presentation.tex

```diff
@@ -0,0 +1,1186 @@
+b_sym = T.vector("b_sym")
+
+# define shared random stream
+trng = T.shared_randomstreams.RandomStreams(1234)
+d=trng.binomial(size=W[1].shape)
+\end{lstlisting}
+\end{frame}
+
+
+\begin{frame}[fragile]
+  \frametitle{Scan Example1: Computing tanh(v.dot(W) + b) * d where d is binomial (2)}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+results, updates = theano.scan(
+    lambda v: T.tanh(T.dot(v, W) + b_sym) * d,
+    sequences=X)
+f = theano.function(inputs=[X, W, b_sym],
+                    outputs=[results],
+                    updates=updates)
+x = np.eye(10, 2, dtype=theano.config.floatX)
+w = np.ones((2, 2), dtype=theano.config.floatX)
+b = np.ones((2), dtype=theano.config.floatX)
+
+print f(x, w, b)
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Scan Example2: Computing pow(A, k)}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+import theano
+import theano.tensor as T
+theano.config.warn.subtensor_merge_bug = False
+
+k = T.iscalar("k")
+A = T.vector("A")
+
+def inner_fct(prior_result, B):
+    return prior_result * B
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Scan Example2: Computing pow(A, k) (2)}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+result, updates = theano.scan(
+    fn=inner_fct,
+    outputs_info=T.ones_like(A),
+    non_sequences=A, n_steps=k)
+
+# Scan provide us with A ** 1 through A ** k.
+# Keep only the last value. Scan optimize memory.
+final = result[-1]
+
+power = theano.function(inputs=[A, k], outputs=final,
+                      updates=updates)
+print power(range(10), 2)
+#[  0.   1.   4.   9.  16.  25.  36.  49.  64.  81.]
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Scan signature}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+result, updates = theano.scan(
+    fn=inner_fct,
+    sequences=[]
+    outputs_info=[T.ones_like(A)],
+    non_sequences=A,
+    n_steps=k)
+\end{lstlisting}
+
+\begin{itemize}
+\item Updates are needed if there are random numbers generated in the inner function
+\begin{itemize}
+\item Pass them to the call theano.function(..., updates=updates)
+\end{itemize}
+\item The inner function of scan takes arguments like this:
+   scan: sequences, outputs\_info, non sequences
+\end{itemize}
+
+\end{frame}
+
+
+\section{RNN}
+\begin{frame}
+  \tableofcontents[currentsection]
+\end{frame}
+
+\begin{frame}
+  \frametitle{Recurrent Neural Network Overview}
+\begin{itemize}
+\item RNN is a class of neural network that allows to work with sequences of variable sizes.
+\item It does so, by reusing weights for each element of the sequence.
```

Comment:
That's not really the defining thing of RNNs. Based on that description, a ConvNet would be an RNN. The defining feature of RNNs is that there is a cycle in the neural graph. This is often implemented as at least one layer having a recurrent connection to itself with a time delay (meaning that the neurons of that layer take as input both the activations of the layer below and their own previous activations